File: apps/syn-cli/src/syn_cli/commands/events.py
"""Event query commands — recent activity, session events, timeline, costs, tools."""

from __future__ import annotations

from typing import Annotated

import typer
from rich.table import Table

from syn_cli._output import console, format_cost, format_duration, format_timestamp
from syn_cli.commands._api_helpers import api_get, api_get_list, build_params

app = typer.Typer(
    name="events",
    help="Query domain events and session observability",
    no_args_is_help=True,
)


@app.command("recent")
def recent(
    limit: Annotated[int, typer.Option(help="Max events to show (max 200)")] = 50,
) -> None:
    """Show recent domain events across all sessions."""
    data = api_get("/events/recent", params={"limit": limit})

    events = data.get("events", [])
    if not events:
        console.print("[dim]No recent events.[/dim]")
        return

    table = Table(title="Recent Events")
    table.add_column("Time")
    table.add_column("Type", style="cyan")
    table.add_column("Session", style="dim")
    table.add_column("Execution", style="dim")

    for ev in events:
        table.add_row(
            format_timestamp(str(ev.get("time", ""))),
            ev.get("event_type", ""),
            (ev.get("session_id") or "")[:12],
            (ev.get("execution_id") or "")[:12],
        )
    console.print(table)
    if data.get("has_more"):
        console.print("[dim]More events available. Use --limit to fetch more.[/dim]")


@app.command("session")
def session_events(
    session_id: Annotated[str, typer.Argument(help="Session ID")],
    event_type: Annotated[
        str | None, typer.Option("--type", "-t", help="Filter by event type")
    ] = None,
    limit: Annotated[int, typer.Option(help="Max events (max 1000)")] = 100,
    offset: Annotated[int, typer.Option(help="Pagination offset")] = 0,
) -> None:
    """List events for a specific session."""
    params = build_params(limit=limit, offset=offset, event_type=event_type)
    data = api_get(f"/events/sessions/{session_id}", params=params)

    events = data.get("events", [])
    if not events:
        console.print(f"[dim]No events for session {session_id}.[/dim]")
        return

    table = Table(title=f"Events: {session_id[:12]}")
    table.add_column("Time")
    table.add_column("Type", style="cyan")
    table.add_column("Phase", style="dim")

    for ev in events:
        table.add_row(
            format_timestamp(str(ev.get("time", ""))),
            ev.get("event_type", ""),
            (ev.get("phase_id") or "")[:12],
        )
    console.print(table)
    if data.get("has_more"):
        console.print(f"[dim]More events available. Use --offset {offset + limit}.[/dim]")


@app.command("timeline")
def timeline(
    session_id: Annotated[str, typer.Argument(help="Session ID")],
    limit: Annotated[int, typer.Option(help="Max entries (max 500)")] = 100,
) -> None:
    """Show a chronological tool-call timeline for a session."""
    entries = api_get_list(f"/events/sessions/{session_id}/timeline", params={"limit": limit})

    if not entries:
        console.print("[dim]No timeline entries.[/dim]")
        return

    table = Table(title=f"Timeline: {session_id[:12]}")
    table.add_column("Time")
    table.add_column("Event", style="cyan")
    table.add_column("Tool")
    table.add_column("Duration", justify="right")
    table.add_column("✓")

    for e in entries:
        dur = e.get("duration_ms")
        success = e.get("success")
        table.add_row(
            format_timestamp(str(e.get("time", ""))),
            e.get("event_type", ""),
            e.get("tool_name") or "—",
            format_duration(dur) if dur is not None else "—",
            "[green]✓[/green]" if success else ("[red]✗[/red]" if success is False else "—"),
        )
    console.print(table)


@app.command("costs")
def session_costs(
    session_id: Annotated[str, typer.Argument(help="Session ID")],
) -> None:
    """Show token usage and cost breakdown for a session."""
    d = api_get(f"/events/sessions/{session_id}/costs")

    console.print(f"[bold]Session costs:[/bold] {d.get('session_id', session_id)}")
    console.print(f"  Input tokens:       {d.get('input_tokens', 0):,}")
    console.print(f"  Output tokens:      {d.get('output_tokens', 0):,}")
    console.print(f"  Total tokens:       {d.get('total_tokens', 0):,}")
    if d.get("cache_creation_tokens"):
        console.print(f"  Cache creation:     {d['cache_creation_tokens']:,}")
    if d.get("cache_read_tokens"):
        console.print(f"  Cache read:         {d['cache_read_tokens']:,}")
    if d.get("estimated_cost_usd") is not None:
        console.print(f"  Estimated cost:     {format_cost(d['estimated_cost_usd'])}")


@app.command("tools")
def session_tools(
    session_id: Annotated[str, typer.Argument(help="Session ID")],
) -> None:
    """Show tool usage summary for a session."""
    tools = api_get_list(f"/events/sessions/{session_id}/tools")

    if not tools:
        console.print("[dim]No tool usage recorded.[/dim]")
        return

    table = Table(title=f"Tool Usage: {session_id[:12]}")
    table.add_column("Tool", style="cyan")
    table.add_column("Calls", justify="right")
    table.add_column("Success", justify="right")
    table.add_column("Errors", justify="right")
    table.add_column("Avg Duration", justify="right")

    for t in sorted(tools, key=lambda x: x.get("call_count", 0), reverse=True):
        avg_ms = t.get("avg_duration_ms")
        table.add_row(
            t.get("tool_name", ""),
            str(t.get("call_count", 0)),
            str(t.get("success_count", 0)),
            str(t.get("error_count", 0)),
            format_duration(avg_ms) if avg_ms else "—",
        )
    console.print(table)

File: apps/syn-cli-node/src/commands/events.ts
/**
 * Event query commands — recent, session, timeline, costs, tools.
 * Port of apps/syn-cli/src/syn_cli/commands/events.py
 */

import { CommandGroup, type CommandDef, type ParsedArgs } from "../framework/command.js";
import { CLIError } from "../framework/errors.js";
import { api, unwrap } from "../client/typed.js";
import type { components } from "../generated/api-types.js";
import { print, printError, printDim } from "../output/console.js";
import { style, BOLD, CYAN, DIM, GREEN, RED } from "../output/ansi.js";
import { formatCost, formatDuration, formatTimestamp } from "../output/format.js";
import { Table } from "../output/table.js";

type EventList = components["schemas"]["EventListResponse"];
type CostSummary = components["schemas"]["syn_api__routes__events__CostSummaryResponse"];
type TimelineEntry = components["schemas"]["TimelineEntryResponse"];
type ToolSummaryItem = components["schemas"]["ToolSummary"];

function reqSessionId(parsed: ParsedArgs): string {
  const id = parsed.positionals[0];
  if (!id) { printError("Missing session-id"); throw new CLIError("Missing argument", 1); }
  return id;
}

const recentCommand: CommandDef = {
  name: "recent",
  description: "Show recent domain events across all sessions",
  options: {
    limit: { type: "string", description: "Max events (max 200)", default: "50" },
    type: { type: "string", short: "t", description: "Filter by event type" },
  },
  handler: async (parsed: ParsedArgs) => {
    const limitStr = (parsed.values["limit"] as string | undefined) ?? "50";
    const eventType = parsed.values["type"] as string | undefined;

    const data: EventList = unwrap(await api.GET("/events/recent", {
      params: {
        query: {
          limit: parseInt(limitStr, 10),
          event_type: eventType ?? null,
        },
      },
    }), "Failed to fetch recent events");

    if (data.events.length === 0) { printDim("No recent events."); return; }

    const table = new Table({ title: "Recent Events" });
    table.addColumn("Time");
    table.addColumn("Type", { style: CYAN });
    table.addColumn("Session", { style: DIM });
    table.addColumn("Execution", { style: DIM });

    for (const ev of data.events) {
      table.addRow(
        formatTimestamp(String(ev.time ?? "")),
        ev.event_type,
        (ev.session_id ?? "").slice(0, 12),
        (ev.execution_id ?? "").slice(0, 12),
      );
    }
    table.print();
    if (data.has_more) printDim("More events available. Use --limit to fetch more.");
  },
};

const sessionEventsCommand: CommandDef = {
  name: "session",
  description: "List events for a specific session",
  args: [{ name: "session-id", description: "Session ID", required: true }],
  options: {
    type: { type: "string", short: "t", description: "Filter by event type" },
    limit: { type: "string", description: "Max events (max 1000)", default: "100" },
    offset: { type: "string", description: "Pagination offset", default: "0" },
  },
  handler: async (parsed: ParsedArgs) => {
    const sid = reqSessionId(parsed);
    const limitStr = (parsed.values["limit"] as string | undefined) ?? "100";
    const offsetStr = (parsed.values["offset"] as string | undefined) ?? "0";
    const eventType = parsed.values["type"] as string | undefined;

    const data: EventList = unwrap(await api.GET("/events/sessions/{session_id}", {
      params: {
        path: { session_id: sid },
        query: {
          limit: parseInt(limitStr, 10),
          offset: parseInt(offsetStr, 10),
          event_type: eventType ?? null,
        },
      },
    }), "Failed to fetch session events");

    if (data.events.length === 0) { printDim(`No events for session ${sid}.`); return; }

    const table = new Table({ title: `Events: ${sid.slice(0, 12)}` });
    table.addColumn("Time");
    table.addColumn("Type", { style: CYAN });
    table.addColumn("Phase", { style: DIM });

    for (const ev of data.events) {
      table.addRow(
        formatTimestamp(String(ev.time ?? "")),
        ev.event_type,
        (ev.phase_id ?? "").slice(0, 12),
      );
    }
    table.print();
    const offset = parseInt(offsetStr, 10);
    const limit = parseInt(limitStr, 10);
    if (data.has_more) printDim(`More events available. Use --offset ${offset + limit}.`);
  },
};

const timelineCommand: CommandDef = {
  name: "timeline",
  description: "Show a chronological tool-call timeline for a session",
  args: [{ name: "session-id", description: "Session ID", required: true }],
  options: { limit: { type: "string", description: "Max entries (max 500)", default: "100" } },
  handler: async (parsed: ParsedArgs) => {
    const sid = reqSessionId(parsed);
    const limitStr = (parsed.values["limit"] as string | undefined) ?? "100";

    const entries: TimelineEntry[] = unwrap(await api.GET("/events/sessions/{session_id}/timeline", {
      params: {
        path: { session_id: sid },
        query: { limit: parseInt(limitStr, 10) },
      },
    }), "Failed to fetch timeline");

    if (entries.length === 0) { printDim("No timeline entries."); return; }

    const table = new Table({ title: `Timeline: ${sid.slice(0, 12)}` });
    table.addColumn("Time");
    table.addColumn("Event", { style: CYAN });
    table.addColumn("Tool");
    table.addColumn("Duration", { align: "right" });
    table.addColumn("\u2713");

    for (const e of entries) {
      table.addRow(
        formatTimestamp(String(e.time ?? "")),
        e.event_type,
        e.tool_name ?? "\u2014",
        e.duration_ms != null ? formatDuration(e.duration_ms) : "\u2014",
        e.success === true ? style("\u2713", GREEN) : (e.success === false ? style("\u2717", RED) : "\u2014"),
      );
    }
    table.print();
  },
};

const costsCommand: CommandDef = {
  name: "costs",
  description: "Show token usage and cost breakdown for a session",
  args: [{ name: "session-id", description: "Session ID", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const sid = reqSessionId(parsed);

    const d: CostSummary = unwrap(await api.GET("/events/sessions/{session_id}/costs", {
      params: { path: { session_id: sid } },
    }), "Failed to fetch session costs");

    print(`${style("Session costs:", BOLD)} ${d.session_id}`);
    print(`  Input tokens:       ${d.input_tokens.toLocaleString()}`);
    print(`  Output tokens:      ${d.output_tokens.toLocaleString()}`);
    print(`  Total tokens:       ${d.total_tokens.toLocaleString()}`);
    if (d.cache_creation_tokens) print(`  Cache creation:     ${d.cache_creation_tokens.toLocaleString()}`);
    if (d.cache_read_tokens) print(`  Cache read:         ${d.cache_read_tokens.toLocaleString()}`);
    if (d.estimated_cost_usd != null)
      print(`  Estimated cost:     ${formatCost(String(d.estimated_cost_usd))}`);
  },
};

const toolsCommand: CommandDef = {
  name: "tools",
  description: "Show tool usage summary for a session",
  args: [{ name: "session-id", description: "Session ID", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const sid = reqSessionId(parsed);

    const tools: ToolSummaryItem[] = unwrap(await api.GET("/events/sessions/{session_id}/tools", {
      params: { path: { session_id: sid } },
    }), "Failed to fetch tool summary");

    if (tools.length === 0) { printDim("No tool usage recorded."); return; }

    const table = new Table({ title: `Tool Usage: ${sid.slice(0, 12)}` });
    table.addColumn("Tool", { style: CYAN });
    table.addColumn("Calls", { align: "right" });
    table.addColumn("Success", { align: "right" });
    table.addColumn("Errors", { align: "right" });
    table.addColumn("Avg Duration", { align: "right" });

    const sorted = [...tools].sort((a, b) => b.call_count - a.call_count);
    for (const t of sorted) {
      table.addRow(
        t.tool_name,
        String(t.call_count),
        String(t.success_count),
        String(t.error_count),
        t.avg_duration_ms ? formatDuration(t.avg_duration_ms) : "\u2014",
      );
    }
    table.print();
  },
};

export const eventsGroup = new CommandGroup("events", "Query domain events and session observability");
eventsGroup.command(recentCommand).command(sessionEventsCommand).command(timelineCommand).command(costsCommand).command(toolsCommand);
